src/librustc/mir/interpret/allocation.rs

```diff
@@ -41,6 +45,411 @@ pub struct Allocation<Tag=(),Extra=()> {
+    pub fn get_bytes_mut(
+        &mut self,
+        cx: &impl HasDataLayout,
+        ptr: Pointer<Tag>,
+        size: Size,
+    ) -> EvalResult<'tcx, &mut [u8]> {
+        assert_ne!(size.bytes(), 0, "0-sized accesses should never even get a `Pointer`");
+        self.check_bounds(cx, ptr, size, true)?;
+
+        self.mark_definedness(ptr, size, true)?;
+        self.clear_relocations(cx, ptr, size)?;
+
+        AllocationExtra::memory_written(self, ptr, size)?;
+
+        assert_eq!(ptr.offset.bytes() as usize as u64, ptr.offset.bytes());
+        assert_eq!(size.bytes() as usize as u64, size.bytes());
+        let offset = ptr.offset.bytes() as usize;
+        Ok(&mut self.bytes[offset..offset + size.bytes() as usize])
+    }
+}
+
+/// Reading and writing
+impl<'tcx, Tag: Copy, Extra: AllocationExtra<Tag>> Allocation<Tag, Extra> {
+    pub fn read_c_str(
+        &self,
+        cx: &impl HasDataLayout,
+        ptr: Pointer<Tag>,
+    ) -> EvalResult<'tcx, &[u8]> {
+        assert_eq!(ptr.offset.bytes() as usize as u64, ptr.offset.bytes());
+        let offset = ptr.offset.bytes() as usize;
+        match self.bytes[offset..].iter().position(|&c| c == 0) {
+            Some(size) => {
+                let p1 = Size::from_bytes((size + 1) as u64);
+                self.check_relocations(cx, ptr, p1)?;
+                self.check_defined(ptr, p1)?;
+                Ok(&self.bytes[offset..offset + size])
+            }
+            None => err!(UnterminatedCString(ptr.erase_tag())),
+        }
+    }
+
+    pub fn check_bytes(
+        &self,
+        cx: &impl HasDataLayout,
+        ptr: Pointer<Tag>,
+        size: Size,
+        allow_ptr_and_undef: bool,
+    ) -> EvalResult<'tcx> {
+        // Check bounds, align and relocations on the edges
+        self.get_bytes_with_undef_and_ptr(cx, ptr, size)?;
+        // Check undef and ptr
+        if !allow_ptr_and_undef {
+            self.check_defined(ptr, size)?;
+            self.check_relocations(cx, ptr, size)?;
+        }
+        Ok(())
+    }
+
+    pub fn write_bytes(
+        &mut self,
+        cx: &impl HasDataLayout,
+        ptr: Pointer<Tag>,
+        src: &[u8],
+    ) -> EvalResult<'tcx> {
+        let bytes = self.get_bytes_mut(cx, ptr, Size::from_bytes(src.len() as u64))?;
+        bytes.clone_from_slice(src);
+        Ok(())
+    }
+
+    pub fn write_repeat(
+        &mut self,
+        cx: &impl HasDataLayout,
+        ptr: Pointer<Tag>,
+        val: u8,
+        count: Size
+    ) -> EvalResult<'tcx> {
+        let bytes = self.get_bytes_mut(cx, ptr, count)?;
+        for b in bytes {
+            *b = val;
+        }
+        Ok(())
+    }
+
+    /// Read a *non-ZST* scalar
+    ///
+    /// zsts can't be read out of two reasons:
+    /// * byteorder cannot work with zero element buffers
+    /// * in oder to obtain a `Pointer` we need to check for ZSTness anyway due to integer pointers
+    ///   being valid for ZSTs
+    pub fn read_scalar(
```

Comment:
Here and for the other high-level methods, it'd be a good idea IMO to add a warning in the doc comment saying that if the caller cares about alignment, they better check it before calling this one.